### PR TITLE
localPointInCylinder now uses correct basis vectors for random point generation

### LINKED ISSUE #25595: Random point generation for hollow cylinders bug
Currently, there is a problem with the random point generation algorithm for hollow cylinders. The current algorithm erroneously generates points within the empty region of the cylinder at radius `0` and as a result, skews the result of the MC absorption correction.
Below is a scatter plot of the sampling done for 10000 points using the current algorithm (Actual behaviour) compared with the algorithm found in Mantid 4.0 (Expected behaviour).
### Expected behavior
Figure_2
### Actual behavior
Figure_1

### Platforms affected
All

### PR DESCRIPTION
**Pull request by aybamidele**
**Description of work.**
I modified localPointInCylinder to use correct basis vectors and included unit tests.

**To test:**
One way to test this is to perform the scatter plots as described in #25595. It is also possible to compare the results of a Montecarlo absorption correction and observe that the results are the same.

Fixes #25595. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Geometry/src/RandomPoint.cpp [hunk 1]
```diff
@@ -100,8 +109,8 @@ Kernel::V3D inHollowCylinder(const detail::ShapeInfo &shapeInfo,
   const double c2 = geometry.radius * geometry.radius;
   const double r{std::sqrt(c1 + (c2 - c1) * r2)};
   const double z{geometry.height * r3};
+  const Kernel::V3D basis1{geometry.axis};
```
Comment (soininen) [inline Framework/Geometry/src/RandomPoint.cpp]: `basis1` here is just an alias to `geometry.axis`. To me, the variable does not make much sense anymore as the basis vector stuff has been moved to `localPointInCylinder()`. I think `basis` could now be removed and `geometry.axis` used directly.

File: Framework/Geometry/src/RandomPoint.cpp [hunk 2]
```diff
@@ -26,11 +26,20 @@ Kernel::V3D localPointInCylinder(const Kernel::V3D &basis,
                                  const Kernel::V3D &alongAxis,
                                  double polarAngle, double radialLength) {
   using boost::math::pow;
```
Comment (soininen) [inline Framework/Geometry/src/RandomPoint.cpp]: Is this needed anymore?

File: Framework/Geometry/src/RandomPoint.cpp [hunk 3]
```diff
@@ -26,11 +26,20 @@ Kernel::V3D localPointInCylinder(const Kernel::V3D &basis,
...
-  Mantid::Kernel::V3D basis2{1., 0., 0.};
-  if (basis.X() != 0. && basis.Z() != 0) {
-    const auto inverseXZSumSq = 1. / (pow<2>(basis.X()) + pow<2>(basis.Z()));
-    basis2.setX(std::sqrt(1. - pow<2>(basis.X()) * inverseXZSumSq));
-    basis2.setZ(basis.X() * std::sqrt(inverseXZSumSq));
+  // Use basis to get a second perpendicular vector to define basis2
+  Mantid::Kernel::V3D basis2;
```
Comment (soininen) [inline Framework/Geometry/src/RandomPoint.cpp]: I think the `Mantid::` part could be removed if one wanted to make this code slightly less wordy.

File: Framework/Geometry/src/RandomPoint.cpp [hunk 4]
```diff
@@ -26,11 +26,20 @@ Kernel::V3D localPointInCylinder(const Kernel::V3D &basis,
...
+  if (basis.X() == 0) {
+    Mantid::Kernel::V3D x{1., 0., 0.};
+    basis2 = x;
```
Comment (soininen) [inline Framework/Geometry/src/RandomPoint.cpp]: Why not just `basis.setX(1.);`? It avoid constructing and copying an entire `V3D`. Ditto for the other cases below.

File: Framework/Geometry/src/RandomPoint.cpp [hunk 5]
```diff
@@ -26,11 +26,20 @@ Kernel::V3D localPointInCylinder(const Kernel::V3D &basis,
...
+  } else if (basis.Y() == 0) {
+    Mantid::Kernel::V3D y{0., 1., 0.};
+    basis2 = y;
+  } else if (basis.Z() == 0) {
+    Mantid::Kernel::V3D z{0., 0., 1.};
+    basis2 = z;
+  } else {
+    Mantid::Kernel::V3D v{-basis.Y(), basis.X(), 0.};
```
Comment (soininen) [inline Framework/Geometry/src/RandomPoint.cpp]: One could avoid the temporary `v` variable by
```c++
basis2.setX(-basis.Y());
basis2.setY(basis.X());
basis2.normalize();
```

File: Framework/Geometry/test/RandomPointTest.h [hunk 6]
```diff
@@ -285,10 +285,20 @@ class RandomPointTest : public CxxTest::TestSuite {
     constexpr V3D alongAxis{0., 0., 1.};
     constexpr V3D basis{1., 1., 1.}; // X and Z elements are not 0 here
 
-    Mantid::Kernel::V3D basis2{1., 0., 0.};
-    const auto inverseXZSumSq = 1. / (pow<2>(basis.X()) + pow<2>(basis.Z()));
-    basis2.setX(std::sqrt(1. - pow<2>(basis.X()) * inverseXZSumSq));
-    basis2.setZ(basis.X() * std::sqrt(inverseXZSumSq));
+    Mantid::Kernel::V3D basis2;
+    if (basis.X() == 0) {
+      Mantid::Kernel::V3D x{1., 0., 0.};
```
Comment (soininen) [inline Framework/Geometry/test/RandomPointTest.h]: Here one could use the `setX()` etc. methods of `V3D`, too.

**Review by soininen — COMMENTED** [review]
It is my second last day at work, so I am not really starting a review here. A few comments on the code, though.

**Comment by martyngigg** [comment]
@aybamidele Am I correct in thinking that to reproduce the original issue then we just need a holly cylinder oriented along `X`?

**Comment by aybamidele** [comment]
Yes. But I would say the problem wasn't really the cylinder's orientation. It was the way the unit vectors used to obtain the local points were defined. They just needed to be mutually independent, and they were not in this case. `basis1` and `basis2` were pointing in the same direction and so `basis3` by its definition (cross product) was 0. And so at angles close to `π/2`  or `2π/3` where the sin function is `0`

**Comment by aybamidele** [comment]
Although this does not explain why there are no intermediate points between 0 and `innerRadius` and how we can span all polar angles with `basis3` equal to 0.

File: Framework/Geometry/src/RandomPoint.cpp [hunk 7]
```diff
@@ -100,8 +105,8 @@ Kernel::V3D inHollowCylinder(const detail::ShapeInfo &shapeInfo,
   const double c2 = geometry.radius * geometry.radius;
   const double r{std::sqrt(c1 + (c2 - c1) * r2)};
   const double z{geometry.height * r3};
+  const Kernel::V3D basis1{geometry.axis};
   const Kernel::V3D alongAxis{geometry.axis * z};
-  const Kernel::V3D &basis1{geometry.axis};
   auto localPoint = localPointInCylinder(basis1, alongAxis, polar, r);
```
Comment (martyngigg) [inline Framework/Geometry/src/RandomPoint.cpp]: Can we call `localPointInCylinder` with `geometry.axis` directly here to avoid the temporary?

**Comment by martyngigg** [comment]
@aybamidele I think the reason there is nothing between `0` and `innerRadius` is that the random number for the radius has a minimum value of `innerRadius` so it only goes to `0` in the cases you mention: `basis3` was always zero if the cylinder axis was `X` so it left `basis2 * std::cos(polarAngle)*radialLength` and this would go to zero when `cos` did.

**Review by martyngigg — APPROVED** [review]
This looks good to me now. I've run the following script:

```python
data = CreateSimulationWorkspace(Instrument='LET', BinParams='0,0.05,25', OutputWorkspace='data', UnitX='Wavelength')
data = ExtractSingleSpectrum(data, 0)
AddSampleLog(data, 'Ei', LogText='2.04',LogType='Number')

ann_temp="""<hollow-cylinder id="A">
<centre-of-bottom-base x="{}" y="0.0" z="0.0" />
<axis x="1.0" y="0.0" z="0.0"/>
<inner-radius val="{}"/>
<outer-radius val="{}"/>
<height val="{}"/>
</hollow-cylinder>"""
sample_formula='Cu2 Se'
#geometry in cm
sample_height=4.0
sample_out_radius=2.0
sample_inn_radius=1.9
sample_density=0.073

shape_xml = ann_temp.format(-0.5*sample_height*0.01,sample_inn_radius*0.01,sample_out_radius*0.01,sample_height*0.01)

#SetSampleMaterial(ws,ChemicalFormula=sample_formula,SampleNumberDensity=sample_density)
SetSample(data, Geometry={'Shape': 'CSG', 'Value': shape_xml},
                           Material={'ChemicalFormula': sample_formula,
                                            'SampleNumberDensity': sample_density})
# Simulating every data point can be slow. Use a smaller set and interpolate
abscor = MonteCarloAbsorption(data, NumberOfWavelengthPoints=20,EventsPerPoint=10000,SparseInstrument=False)
```
and confirmed that this now gives the same answer as v4.0 and fixes the problem on with master with reporting higher absorption than was actually the case.

**Pull request merged**